### PR TITLE
Export `run` function as a streamlined but flexible main entrypoint 

### PR DESCRIPTION
**Pull request by timriley**
Introduce a new `assets.run` entrypoint function that is:

- Streamlined enough for us to generate into every new Hanami app
- But flexible enough for users to provide arbitrary esbuild customisations

Here's what a default `config/assets.mjs` file would look like using this API:

```js
import * as assets from "hanami-assets";

await assets.run();
```

And here's how it would look to add a custom esbuild plugin:

```js
import * as assets from "hanami-assets";
import postcss from "esbuild-postcss";

await assets.run({
  esbuildOptionsFn: (args, esbuildOptions) => {
    const plugins = [...esbuildOptions.plugins, postcss()];

    return {
      ...esbuildOptions,
      plugins,
    };
  },
});
```

I've tested this exact config above in a real Hanami app and can confirm it actually works in terms of activating esbuild plugins.

---

In terms of code changes here, they are fairly wide-ranging. Sorry for the messy diff.

Let me guide you through the new landscape, however:

* **index.ts**: defines the `run` function that serves as the main entry point, as shown above in those `config/assets.mjs` examples. This accepts an object of options so that they can all be optional.
* **args.ts**: a new file that defines a proper interface and return object for the args that we parse when invoking `node config/assets.mjs` inside Hanami apps. This makes sure t things are nicely typed when passing an `esbuildOptionsFn` per the second example above.
* **esbuild.ts**: defines our standard esbuild options, as used by the run function above. Most of the code here has been simply moved into a new location, otherwise left unchanged, aside from the new `buildOptions` and `watchOptions` functions that return the combined set of options for the two modes that we use.
* **esbuild-plugin.ts**: defines our esbuild plugin. I've barely touched this file, except for a few light renames of its exports.

That really sums it up! If you spend most of your time checking out `index.ts`, `args.ts` and skimming `esbuild.ts`, you'll be up to speed with the changes.

---

There are a few other things I took care of while I was here:

- Removed the config indicating that this package exposes an executable. Running via a ` config/assets.mjs` script is now the one and only way to use this package.
- Use [Prettier](https://prettier.io) for JS file linting/auto-formatting. This introduced a fair bit of churn, but I needed it because I wasn't sure how I should be formatting my TypeScript. 😆 And now's as good a time as any to get this into place.
- Update TypeScript `compilerOptions` to use `"module": "NodeNext"`. Per their [relevant docs](https://www.typescriptlang.org/tsconfig#module):
  > You very likely want `"nodenext"` for modern node projects.

---

For full context of how this will be used within a full Hanami app, see https://github.com/hanami/cli/pull/109.

### CODE REVIEW

File: src/esbuild.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      modifiedPath = name;
+    }
+
+    result[modifiedPath] = entryPoint;
+  });
+
+  return result;
+};
+
+// TODO: feels like this really should be passed a root too, to become the cwd for globSync
```
Comment (timriley) [inline src/esbuild.ts:60]: This was the other TODO I added here. The one other place in this file where we were doing file operations without providing an explicit root.

File: src/esbuild.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    });
+
+    return filteredDirs.map((dir) => path.join(dir, "*"));
+  } catch (err) {
+    console.error("Error listing external directories:", err);
+    return [];
+  }
+};
+
+// TODO: reuse the logic between these two methods below
```
Comment (timriley) [inline src/esbuild.ts:83]: I also added this TODO, but I'd rather leave it for another PR to refactor these. I left these lists of options mostly unchanged for the sake of continuity.

**Comment by timriley** [comment]
Lastly, apologies for the messy commit history here. This took a few twists and turns (and some of the usual "Tim figures out how to JavaScript again") before I got everything settled.

This is going to be squashed on merge anyway, and I'll make sure there's a good single commit message then. In the meantime, please see my overview in the PR description for the best way of inspecting the changes in this PR.

File: src/esbuild.ts (hunk 1 above)
Comment (jodosha) [reply src/esbuild.ts:60]: @timriley IIRC, by adding the absolute path, Esbuild will report the absolute path in the metadata, so it's a bit cumbersome to work lately with them in the context of manifest generation.

**Review by jodosha — APPROVED** [review]
@timriley On the other day you were "unreasonably excited" about these changes. Now I understand why: they are magnificent! They're taking this code base to the next level both in what they offer to end users, but also in quality of the internal code. Well done! 👏 

Please check my inline comment about `package.json`. 🙂

File: src/esbuild.ts (hunk 1 above)
Comment (timriley) [reply src/esbuild.ts:60]: Ah, this is useful to know, thanks @jodosha!

The only downside of the current arrangement is that it's kind of contingent of the cwd also equalling the app root. In 99% of usage, I don't expect these things will diverge, but the fact that we're allowing a `root:` to be provided to `assets.run` means that some of these globs that reply on the cwd might start to fail.

However, I'm comfortable enough with us leaving this as a known gap for the moment. We can come back to solidify this aspect later.

And perhaps we don't document the `assets.run` `root:` option as "public" for now, too.

File: package.json [hunk 3]
```diff
@@ -37,18 +38,13 @@
     "@types/react": "^18.2.16",
     "@types/react-dom": "^18.2.7",
     "jest": "^29.6.1",
+    "prettier": "3.0.3",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "ts-jest": "^29.1.1",
+    "ts-node": "^10.9.1",
```
[the remark this answers is not in the document]
Comment (timriley) [reply package.json:45]: Thanks for pointing this out! I think I added ts-node when I was struggling to get everything to work again after temporarily breaking it. Your comment made me realise that I don't think it's necessary for anything at the moment, so I've removed it again for now.

File: package.json [hunk 4]
```diff
@@ -1,12 +1,13 @@
 {
   "name": "hanami-assets",
   "version": "2.1.0-beta.2",
-  "description": "Hanami Assets management via Esbuild",
+  "type": "module",
+  "description": "Hanami assets management via Esbuild",
   "bin": {
```
[the remark this answers is not in the document]
Comment (timriley) [reply package.json:6]: Ooh yes, I thought I removed it and even mentioned in the PR description, but clearly did not 😆
